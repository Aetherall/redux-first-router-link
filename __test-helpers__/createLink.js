import React from 'react'
import renderer from 'react-test-renderer'
import { createStore, applyMiddleware } from 'redux'
import { Provider } from 'react-redux'
import createHistory from 'history/createMemoryHistory'
import { connectTypes } from 'pure-redux-router'

import Link from '../src/Link'


export default (props) => {
  const routesMap = {
    FIRST: '/first',
    SECOND: '/second/:param',
  }

  const history = createHistory({
    initialEntries: ['/'],
    initialIndex: 0,
    keyLength: 6,
  })

  const { middleware, reducer: locationReducer } = connectTypes(history, routesMap)

  const middlewares = applyMiddleware(middleware)

  const reducer = (state = {}, action = {}) => ({
    location: locationReducer(state.location, action),
  })

  const store = createStore(reducer, undefined, middlewares)

  const link = <Link {...props} />
  const component = renderer.create(<Provider store={store}>{link}</Provider>)

  return {
    tree: component.toJSON(),
    store,
  }
}


export const event = { preventDefault: () => undefined }