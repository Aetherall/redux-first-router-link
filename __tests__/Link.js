import createLink, { event } from '../__test-helpers__/createLink'


it('ON_CLICK: dispatches location-aware action', () => {
  const { tree, store } = createLink({ href: '/first', children: 'CLICK ME', target: '_self' })

  console.log(tree)
  expect(tree).toMatchSnapshot()

  console.log(store.getState())

  tree.props.onClick(event)

  const { location } = store.getState()
  console.log(location)

  expect(location.pathname).toEqual('/first')
  expect(location.type).toEqual('FIRST')
})


it('ON_CLICK: does NOT dispatch if shouldDispatch === false', () => {
  const { tree, store } = createLink({ href: '/first', shouldDispatch: false })

  expect(tree).toMatchSnapshot()

  tree.props.onClick()

  const { location } = store.getState()
  console.log(location)

  expect(location.pathname).not.toEqual('/first')
  expect(location.type).not.toEqual('FIRST')
  expect(location.pathname).toEqual('/')
})


it('ON_CLICK: does NOT dispatch if onPress returns false', () => {
  const { tree, store } = createLink({ href: '/first', onPress: () => false })

  expect(tree).toMatchSnapshot()

  tree.props.onClick()

  const { location } = store.getState()
  console.log(location)

  expect(location.pathname).not.toEqual('/first')
  expect(location.type).not.toEqual('FIRST')
  expect(location.pathname).toEqual('/')
})


it('ON_CLICK: DOES dispatch if onPress returns true', () => {
  const { tree, store } = createLink({ href: '/first', onPress: () => true })

  expect(tree).toMatchSnapshot()

  tree.props.onClick()

  const { location } = store.getState()
  console.log(location)

  expect(location.pathname).toEqual('/first')
  expect(location.type).toEqual('FIRST')
})


it('ON_CLICK: DOES dispatch if onPress returns undefined', () => {
  const { tree, store } = createLink({ href: '/first', onPress: () => undefined })

  expect(tree).toMatchSnapshot()

  tree.props.onClick()

  const { location } = store.getState()
  console.log(location)

  expect(location.pathname).toEqual('/first')
  expect(location.type).toEqual('FIRST')
})


it('ON_MOUSE_DOWN: dispatches action onMouseDown if down === true', () => {
  const { tree, store } = createLink({ href: '/first', down: true })

  console.log(tree)
  expect(tree).toMatchSnapshot()

  tree.props.onClick(event) // e.preventDefault() called and nothing dispatched + linking prevented

  let { location } = store.getState()
  expect(location.pathname).toEqual('/')

  tree.props.onMouseDown()

  location = store.getState().location
  console.log(location)

  expect(location.pathname).toEqual('/first')
  expect(location.type).toEqual('FIRST')
})


it('ON_TOUCH_START: dispatches action onTouchStart if down === true', () => {
  const { tree, store } = createLink({ href: '/first', down: true })

  console.log(tree)
  expect(tree).toMatchSnapshot()

  tree.props.onTouchStart()

  const { location } = store.getState()
  console.log(location)

  expect(location.pathname).toEqual('/first')
  expect(location.type).toEqual('FIRST')
})

