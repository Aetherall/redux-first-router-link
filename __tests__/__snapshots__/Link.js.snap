exports[`test ON_CLICK: DOES dispatch if onPress returns true 1`] = `
<a
  href="/first"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test ON_CLICK: DOES dispatch if onPress returns true 2`] = `
<a
  href="/first"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test ON_CLICK: DOES dispatch if onPress returns undefined 1`] = `
<a
  href="/first"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test ON_CLICK: dispatches location-aware action 1`] = `
<a
  href="/first"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target="_self">
  CLICK ME
</a>
`;

exports[`test ON_CLICK: dispatches location-aware action 2`] = `
<a
  href="/first"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target="_self">
  CLICK ME
</a>
`;

exports[`test ON_CLICK: does NOT dispatch if onPress returns false 1`] = `
<a
  href="/first"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test ON_CLICK: does NOT dispatch if onPress returns false 2`] = `
<a
  href="/first"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test ON_CLICK: does NOT dispatch if shouldDispatch === false 1`] = `
<a
  href="/first"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test ON_MOUSE_DOWN: dispatches action onMouseDown if down === true 1`] = `
<a
  href="/first"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  target={undefined} />
`;

exports[`test ON_TOUCH_START: dispatches action onTouchStart if down === true 1`] = `
<a
  href="/first"
  onClick={[Function]}
  onMouseDown={[Function]}
  onTouchStart={[Function]}
  target={undefined} />
`;

exports[`test converts href as action object to path 1`] = `
<a
  href="/second/bar"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test converts href as action object to path 2`] = `
<a
  href="/second/bar"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test converts href as array of strings to path 1`] = `
<a
  href="/second/bar"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test converts href as array of strings to path 2`] = `
<a
  href="/second/bar"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test converts href as non-matched action object to "#" for path 1`] = `
<a
  href="#"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test converts href as non-matched action object to "#" for path 2`] = `
<a
  href="#"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test converts invalid hre to "#" for path 1`] = `
<a
  href="#"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;

exports[`test converts invalid href to "#" for path 1`] = `
<a
  href="#"
  onClick={[Function]}
  onMouseDown={false}
  onTouchStart={false}
  target={undefined} />
`;
